sound/soc/sof/control.c

```diff
@@ -180,22 +186,26 @@ int snd_sof_enum_put(struct snd_kcontrol *kcontrol,
 	struct snd_sof_dev *sdev = scontrol->sdev;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
 	unsigned int i, channels = scontrol->num_channels;
+	int change = 0;
+	u32 value;
 
 	/* update each channel */
 	for (i = 0; i < channels; i++) {
-		cdata->chanv[i].value = ucontrol->value.enumerated.item[i];
+		value = ucontrol->value.enumerated.item[i];
+		change |= (value != cdata->chanv[i].value);
 		cdata->chanv[i].channel = i;
+		cdata->chanv[i].value = value;
 	}
 
 	/* notify DSP of enum updates */
-	if (pm_runtime_active(sdev->dev))
+	if (change && pm_runtime_active(sdev->dev))
```

Comment:
@dragosht wondering if we can have a case where the values maintained by the driver did not change, so we don't update the DSP values with an IPC, but the values at the DSP levels are initially set to a different default, so the DSP will *never* see the new value.
In other words, could we have a cache effect with no write-back?

Reply:
@plbossart Now that you mentioned it, I can think of a few things ...
- I see none of the put() handlers are currently checking the return of `snd_sof_ipc_set_get_comp_data()` when sending out the IPC to the DSP. If (hypothetically) something is to go wrong in there, the ALSA kcontrol and the DSP get out of sync. And with my patch on top it would even report the change as successful to the userspace if those values were different in the first place. I could put together an additional patch doing the error handling (return an error in put() while also revert the ucontrol value on error) if you find this useful.
- Maybe we should send the IPC regardless of whether something actually changed - haven't looked in depth, but I assume things can go wrong on the DSP side as well

Reply:
@juimonen @ranj063 we need you to chime in on this this one, thanks!

Reply:
Now thinking of this, maybe just send it anyway to dsp, do the error check, return boolean for mixer sliders to update. @ranj063 ?

Reply:
@juimonen @plbossart we had this discussion back when we refactored the set_get_comp_data() too. 

If the DSP is in D0 when the kcontrol changes happen, then the IPC is sent and yes it can fail. In this case, reporting an error to the userspace might be helpful. @singalsu didnt you have a patch for this? or did we decide that we just store the value anyway and retry sending it to the DSP after the next runtime suspend/resume cycle?

Reply:
@ranj063 @juimonen One more thought: if put() is to return an error on that ipc failure, wouldn't it make sense for those `chanv` pairs to get reverted to their previous value?

Reply:
yes I think so. sorry this is little bit confusing as none of us seem to remember our previous discussions about this... :)
but all together:
- let's send the message to dsp anyway
- I think with failure the revert would be fine (I have to say that if at this point there's error, something is really, really badly wrong, but anyway it is a theoretical possibility)
- return the boolean so that mixer sliders change

So with your pr the main thing is to return something else than always zero, we can change later the functionality of sending/not sending always the value to dsp (so for now for safety, let's just send it).